--- qemu-2.3.0-branch/tcg/tcg.h.orig	2017-08-18 10:12:58.561955139 +0900
+++ qemu-2.3.0-branch/tcg/tcg.h	2017-08-18 10:12:44.066164190 +0900
@@ -29,6 +29,15 @@
 #include "qemu/bitops.h"
 #include "tcg-target.h"
 
+/* Constants used for Eclipser instrumentation. Specifies whether an instruction
+ * is target of Eclipser instrumentation (i.e. decides whether to insert code
+ * that records Eflags register value).
+ */
+#define ECLIPSER_CMP_EQUALITY 0
+#define ECLIPSER_CMP_SIZE_SIGNED 1
+#define ECLIPSER_CMP_SIZE_UNSIGNED 2
+#define ECLIPSER_IGNORE 3
+
 /* Default target word size to pointer size.  */
 #ifndef TCG_TARGET_REG_BITS
 # if UINTPTR_MAX == UINT32_MAX
