--- qemu-2.10.0-bbcount/Makefile.target.orig	2020-09-29 08:25:21.155543947 -0700
+++ qemu-2.10.0-bbcount/Makefile.target	2020-09-29 08:22:26.302363966 -0700
@@ -92,6 +92,7 @@
 #########################################################
 # cpu emulator library
 obj-y += exec.o
+obj-y += eclipser.o
 obj-y += accel/
 obj-$(CONFIG_TCG) += tcg/tcg.o tcg/tcg-op.o tcg/optimize.o
 obj-$(CONFIG_TCG) += tcg/tcg-common.o tcg/tcg-runtime.o
