--- qemu-2.3.0-coverage/cpu-exec.c.orig	2020-09-21 23:23:17.517246984 -0700
+++ qemu-2.3.0-coverage/cpu-exec.c	2020-09-21 23:22:27.909433624 -0700
@@ -28,6 +28,13 @@
 #include "exec/memory-internal.h"
 #include "qemu/rcu.h"
 
+extern abi_ulong chatkey_entry_point; /* ELF entry point (_start) */
+extern void chatkey_setup_before_forkserver(void);
+extern void chatkey_setup_after_forkserver(void);
+extern void chatkey_log_bb(abi_ulong addr, abi_ulong callsite);
+
+#include "afl-qemu-cpu-inl.h"
+
 /* -icount align implementation. */
 
 typedef struct SyncClocks {
@@ -296,8 +303,11 @@
     }
  not_found:
    /* if no translated code available, then translate it now */
+
     tb = tb_gen_code(cpu, pc, cs_base, flags, 0);
 
+    afl_request_tsl(pc, cs_base, flags);
+
  found:
     /* Move the last found TB to the head of the list */
     if (likely(*ptb1)) {
@@ -492,6 +502,15 @@
                     next_tb = 0;
                     tcg_ctx.tb_ctx.tb_invalidated_flag = 0;
                 }
+
+                if(tb->pc == chatkey_entry_point) {
+                  chatkey_setup_before_forkserver();
+                  afl_forkserver(env);
+                  chatkey_setup_after_forkserver();
+                }
+
+                chatkey_log_bb(tb->pc, 0);
+
                 if (qemu_loglevel_mask(CPU_LOG_EXEC)) {
                     qemu_log("Trace %p [" TARGET_FMT_lx "] %s\n",
                              tb->tc_ptr, tb->pc, lookup_symbol(tb->pc));
