--- qemu-2.3.0-syscall/cpu-exec.c.orig	2018-05-24 11:30:49.615769240 +0900
+++ qemu-2.3.0-syscall/cpu-exec.c	2018-05-23 17:47:35.161933753 +0900
@@ -28,6 +28,9 @@
 #include "exec/memory-internal.h"
 #include "qemu/rcu.h"
 
+extern abi_ulong chatkey_entry_point;
+extern void chatkey_setup(void);
+
 /* -icount align implementation. */
 
 typedef struct SyncClocks {
@@ -492,6 +495,11 @@
                     next_tb = 0;
                     tcg_ctx.tb_ctx.tb_invalidated_flag = 0;
                 }
+
+                if(tb->pc == chatkey_entry_point) {
+                  chatkey_setup();
+                }
+
                 if (qemu_loglevel_mask(CPU_LOG_EXEC)) {
                     qemu_log("Trace %p [" TARGET_FMT_lx "] %s\n",
                              tb->tc_ptr, tb->pc, lookup_symbol(tb->pc));
