--- qemu-2.3.0-feedback/Makefile.target.orig	2017-04-13 20:44:12.868483773 +0900
+++ qemu-2.3.0-feedback/Makefile.target	2017-04-10 16:46:23.364441824 +0900
@@ -83,7 +83,7 @@
 #########################################################
 # cpu emulator library
 obj-y = exec.o translate-all.o cpu-exec.o
-obj-y += tcg/tcg.o tcg/tcg-op.o tcg/optimize.o
+obj-y += tcg/tcg.o tcg/tcg-op.o tcg/optimize.o tcg/chatkey.o
 obj-$(CONFIG_TCG_INTERPRETER) += tci.o
 obj-$(CONFIG_TCG_INTERPRETER) += disas/tci.o
 obj-y += fpu/softfloat.o
