--- qemu-2.3.0-bbcount/Makefile.target.orig	2018-05-24 15:42:00.432802634 +0900
+++ qemu-2.3.0-bbcount/Makefile.target	2018-05-24 15:19:52.472308828 +0900
@@ -11,7 +11,7 @@
 endif
 QEMU_CFLAGS += -I.. -I$(SRC_PATH)/target-$(TARGET_BASE_ARCH) -DNEED_CPU_H
 
-QEMU_CFLAGS+=-I$(SRC_PATH)/include
+QEMU_CFLAGS+=-I$(SRC_PATH)/include -I$(SRC_PATH)/../../sparsehash/sparsehash-2.0.3/src -I$(SRC_PATH)/../../sparsehash/build/src
 
 ifdef CONFIG_USER_ONLY
 # user emulator name
